cmd/isp/src/protocol.rs

```diff
@@ -0,0 +1,544 @@
+}
+
+#[repr(C)]
+#[derive(Debug, AsBytes, FromBytes)]
+pub struct PingResponse {
+    header: PacketHeader,
+    protocol_bugfix: u8,
+    protocol_minor: u8,
+    protocol_major: u8,
+    protocol_name: u8,
+    options_low: u8,
+    options_high: u8,
+    crc16_low: u8,
+    crc16_high: u8,
+}
+
+#[repr(C)]
+#[derive(Debug, AsBytes, FromBytes)]
+pub struct FramingPacket {
+    header: PacketHeader,
+    length_low: u8,
+    length_high: u8,
+    crc16_low: u8,
+    crc16_high: u8,
+}
+
+impl FramingPacket {
+    fn new(ptype: PacketType) -> FramingPacket {
+        FramingPacket {
+            header: PacketHeader::new(ptype),
+            length_low: 0,
+            length_high: 0,
+            crc16_low: 0,
+            crc16_high: 0,
+        }
+    }
+}
+
+#[derive(Debug, Default, AsBytes, FromBytes)]
+#[repr(C)]
+pub struct RawCommand {
+    tag: u8,
+    flags: u8,
+    reserved: u8,
+    parameter_count: u8,
+}
+
+impl RawCommand {
+    fn new(c: CommandTag, count: usize) -> RawCommand {
+        RawCommand {
+            tag: c as u8,
+            flags: 0,
+            reserved: 0,
+            parameter_count: count as u8,
+        }
+    }
+}
+
+// Command packets can take a variable number
+// of arguments. This is unfortunately a pain to serialize
+// in a structure. So we cheat and later append the
+// arguments manually
+#[derive(Debug, AsBytes)]
+#[repr(C)]
+pub struct VariablePacket {
+    packet: FramingPacket,
+    raw_command: RawCommand,
+}
+
+pub struct CommandPacket {
+    packet: VariablePacket,
+    params: Vec<u32>,
+}
+
+impl CommandPacket {
+    fn new_command(c: CommandTag, args: Vec<u32>) -> Result<CommandPacket> {
+        let mut v = VariablePacket {
+            packet: FramingPacket::new(PacketType::Command),
+            raw_command: RawCommand::new(c, args.len()),
+        };
+
+        let arg_bytes = args.len() * 4;
+        // Total length of the command packet. the 4 bytes are for
+        // the fixed fields
+        let len: u16 = (4 + arg_bytes) as u16;
+
+        v.packet.length_low = (len & 0xFF) as u8;
+        v.packet.length_high = ((len >> 8) & 0xff) as u8;
+
+        let mut crc = CRCu16::crc16xmodem();
+
+        let bytes = v.as_bytes();
+
+        // CRC over everything except the CRC field, this includes the framing
+        // header as well as the rest of the argument
+        crc.digest(&bytes[..0x4]);
+        crc.digest(&bytes[0x6..]);
+
+        for e in args.iter() {
+            crc.digest(&e.to_le_bytes());
+        }
+
+        let digest = crc.get_crc();
+
+        v.packet.crc16_low = (digest & 0xff) as u8;
+        v.packet.crc16_high = ((digest >> 8) & 0xff) as u8;
+
+        Ok(CommandPacket { packet: v, params: args })
+    }
+
+    fn to_bytes(&self) -> Result<Vec<u8>> {
+        let mut v = Vec::new();
+
+        v.extend_from_slice(&self.packet.as_bytes());
+
+        for e in self.params.iter() {
+            v.extend_from_slice(&e.to_le_bytes());
+        }
```

Comment:
If you want to get fancy
```suggestion
        v.extend(self.params.iter().flat_map(|e| e.to_le_bytes()));
```